source/Mlos.Python/mlos/Grpc/BayesianOptimizerProxy.py

```diff
@@ -102,17 +108,19 @@ def register(self, feature_values_pandas_frame, target_values_pandas_frame):
         self._optimizer_stub.RegisterObservations(register_request)
 
     @trace()
-    def get_all_observations(self) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    def get_all_observations(self) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:

```

Comment:
This is probably not the PR to do this, but a thought for later perhaps: would this be a good place for named tuples?
After some point returning a list of data potentially becomes a readability issue.

Reply:
Agreed.

We spoke about this with Andreas a couple of days ago. I think a full class would be best here. For once, named tuples are immutable and they can only be composed of immutable objects, so they cannot hold dataframes. Additionally, we might want to carry some metadata along too (i.e. the OptimizationProblem with the schema definitions for parameters, context, and objectives).

But like you said, it's a bigger change and we agreed to punt it to some later PR.

Reply:
Also, if we keep all three dataframes in an object, we can frequently assert that they have the same length, which should help protect data integrity.

Reply:
Hm you can have named tuples of mutable data I think, but I agree a full object might be nicer.